toolchain/lexer/tokenized_buffer.h

```diff
@@ -23,161 +23,165 @@
-  // its tokens.
-  class RealLiteralValue : public Printable<RealLiteralValue> {
-   public:
-    // The mantissa, represented as an unsigned integer.
-    [[nodiscard]] auto Mantissa() const -> const llvm::APInt& {
-      return buffer_->literal_int_storage_[literal_index_];
-    }
-    // The exponent, represented as a signed integer.
-    [[nodiscard]] auto Exponent() const -> const llvm::APInt& {
-      return buffer_->literal_int_storage_[literal_index_ + 1];
-    }
-    // If false, the value is mantissa * 2^exponent.
-    // If true, the value is mantissa * 10^exponent.
-    [[nodiscard]] auto IsDecimal() const -> bool { return is_decimal_; }
-
-    auto Print(llvm::raw_ostream& output_stream) const -> void {
-      output_stream << Mantissa() << "*" << (is_decimal_ ? "10" : "2") << "^"
-                    << Exponent();
-    }
+  Token token_;
+};
 
-   private:
-    friend class TokenizedBuffer;
+// The value of a real literal.
+//
+// This is either a dyadic fraction (mantissa * 2^exponent) or a decadic
+// fraction (mantissa * 10^exponent).
+//
+// The `TokenizedBuffer` must outlive any `RealLiteralValue`s referring to
+// its tokens.
+class RealLiteralValue : public Printable<RealLiteralValue> {
+ public:
+  // The mantissa, represented as an unsigned integer.
+  [[nodiscard]] auto Mantissa() const -> const llvm::APInt& {
+    return (*literal_int_storage_)[literal_index_];
+  }
+  // The exponent, represented as a signed integer.
+  [[nodiscard]] auto Exponent() const -> const llvm::APInt& {
+    return (*literal_int_storage_)[literal_index_ + 1];
+  }
+  // If false, the value is mantissa * 2^exponent.
+  // If true, the value is mantissa * 10^exponent.
+  [[nodiscard]] auto IsDecimal() const -> bool { return is_decimal_; }
 
-    RealLiteralValue(const TokenizedBuffer* buffer, int32_t literal_index,
-                     bool is_decimal)
-        : buffer_(buffer),
-          literal_index_(literal_index),
-          is_decimal_(is_decimal) {}
+  auto Print(llvm::raw_ostream& output_stream) const -> void {
+    output_stream << Mantissa() << "*" << (is_decimal_ ? "10" : "2") << "^"
+                  << Exponent();
+  }
 
-    const TokenizedBuffer* buffer_;
-    int32_t literal_index_;
-    bool is_decimal_;
-  };
+ private:
+  friend class TokenizedBuffer;
 
-  // A diagnostic location translator that maps token locations into source
-  // buffer locations.
-  class TokenLocationTranslator : public DiagnosticLocationTranslator<Token> {
-   public:
-    explicit TokenLocationTranslator(const TokenizedBuffer* buffer)
-        : buffer_(buffer) {}
+  RealLiteralValue(const llvm::SmallVector<llvm::APInt>* literal_int_storage,
+                   int32_t literal_index, bool is_decimal)
+      : literal_int_storage_(literal_int_storage),
+        literal_index_(literal_index),
+        is_decimal_(is_decimal) {}
+
+  const llvm::SmallVector<llvm::APInt>* literal_int_storage_;
+  int32_t literal_index_;
+  bool is_decimal_;
```

Comment:
Can we store an `ArrayRef<llvm::APInt>` or maybe a pair of `llvm::APInt*`s here, pointing directly to the storage for this value, rather than a vector pointer plus an index?

I think invalidation through reallocation of the `storage` array is not a problem here, because `RealLiteralValue`s are only created after we've already finished forming the `TokenizedBuffer`, at which point it's immutable.

Reply:
Done. Switched to a struct since we're looking at this, hoping that's okay based on discussion (it falls out if it's just a pair of llvm::APInt*)